rasa/core/evaluation/markers.py

```diff
@@ -0,0 +1,120 @@
+import os
+from pathlib import Path
+from typing import (
+    Dict,
+    Text,
+    Union,
+)
+
+import rasa.shared.core.constants
+from rasa.shared.exceptions import RasaException, YamlException, YamlSyntaxException
+import rasa.shared.nlu.constants
+import rasa.shared.utils.validation
+import rasa.shared.utils.io
+import rasa.shared.utils.common
+
+
+class InvalidMarkersConfig(RasaException):
+    """Exception that can be raised when markers config is not valid."""
+
+
+class MarkerConfig:
+    """A class that represents the markers config.
+
+    A markers config contains the markers and the conditions for when they apply.
+    The class reads the config, validates the schema, and validates the conditions.
+    """
+
+    @classmethod
+    def empty_config(cls) -> Dict:
+        """Returns an empty config file."""
+        return {}  # currently represented as a simple dict. Might change later.
+
+    @classmethod
+    def load_config_from_path(cls, path: Union[Text, Path]) -> Dict:
+        """Loads the config from a file or directory."""
+        path = os.path.abspath(path)
+        if os.path.isfile(path):
+            config = cls.from_file(path)
+        elif os.path.isdir(path):
+            config = cls.from_directory(path)
+        else:
+            raise InvalidMarkersConfig(
+                "Failed to load markers configuration from '{}'. "
+                "File not found!".format(os.path.abspath(path))
+            )
+        return config
+
+    @classmethod
+    def from_file(cls, path: Text) -> Dict:
+        """Loads the config from a YAML file."""
+        return cls.from_yaml(rasa.shared.utils.io.read_file(path), path)
+
+    @classmethod
+    def from_yaml(cls, yaml: Text, original_filename: Text = "") -> Dict:
+        """Loads the config from YAML text after validating it."""
+        try:
+            # TODO implement and call the yaml schema validation
+            #  rasa.shared.utils.validation.validate_yaml_schema(yaml,
+            #  MAKERS_SCHEMA_FILE)
+
+            config = rasa.shared.utils.io.read_yaml(yaml)
+            return config
+        except YamlException as e:
+            e.filename = original_filename
+            raise e
+
+    @classmethod
+    def from_directory(cls, path: Text) -> Dict:
+        """Loads and appends multiple configs from a directory tree."""
+        config = cls.empty_config()
+        for root, _, files in os.walk(path, followlinks=True):
+            for file in files:
```

Comment:
Code climate says this is too cognitively complex: https://codeclimate.com/github/RasaHQ/rasa/pull/9729
But I think this is a pretty standard way of generating the filenames in a directory. Do I need to simplify it? 🤔

Reply:
The code climate thing doesn't look like a compulsory test, so I could ignore it in this instance since this is a standard piece of code.

Reply:
Yeah, it's caused me bother before - I think there's plans to switch to a different tool for coverage etc